src/cli/commands/check.js

```diff
@@ -39,108 +40,90 @@ export async function verifyTreeCheck(
+        if (!isLinkedDepencency) {
+          dependenciesToCheckVersion.push({
+            name,
+            key: name,
+            searchPath: [rootModulesFolder],
+            version,
+          });
+        }
       }
-      dependenciesToCheckVersion.push({
-        name,
-        originalKey: name,
-        parentCwd: registry.cwd,
-        version,
-      });
     }
   }
-  if (rootManifest.devDependencies && !config.production) {
-    for (const name in rootManifest.devDependencies) {
-      const version = rootManifest.devDependencies[name];
-      // skip linked dependencies
-      const isLinkedDepencency = /^link:/i.test(version) || (/^file:/i.test(version) && config.linkFileDependencies);
-      if (isLinkedDepencency) {
-        continue;
-      }
-      dependenciesToCheckVersion.push({
-        name,
-        originalKey: name,
-        parentCwd: registry.cwd,
-        version,
-      });
-    }
+  pushDependencies(rootManifest.dependencies);
+  if (!config.production) {
+    pushDependencies(rootManifest.devDependencies);
   }
 
-  const locationsVisited: Set<string> = new Set();
+  const missing = false;
+  const anyVersion = true;
+  const pkgVersions: Map<string, string | boolean> = new Map();
   while (dependenciesToCheckVersion.length) {
     const dep = dependenciesToCheckVersion.shift();
-    const manifestLoc = path.join(dep.parentCwd, registry.folder, dep.name);
-    if (locationsVisited.has(manifestLoc + `@${dep.version}`)) {
-      continue;
-    }
-    locationsVisited.add(manifestLoc + `@${dep.version}`);
-    if (!await fs.exists(manifestLoc)) {
-      reportError('packageNotInstalled', `${dep.originalKey}`);
-      continue;
-    }
-    if (!await fs.exists(path.join(manifestLoc, 'package.json'))) {
-      continue;
-    }
-    const pkg = await config.readManifest(manifestLoc, registryName);
-    if (
-      semver.validRange(dep.version, config.looseSemver) &&
-      !semver.satisfies(pkg.version, dep.version, config.looseSemver)
-    ) {
-      reportError('packageWrongVersion', dep.originalKey, dep.version, pkg.version);
-      continue;
+    const searchPath = dep.searchPath;
+    let version: string | boolean = false;
+    let dependencies: ?Dependencies;
+
+    while (searchPath.length) {
+      const manifestLoc = path.join(...searchPath, dep.name);
+
+      if (pkgVersions.has(manifestLoc)) {
+        version = pkgVersions.get(manifestLoc) || missing;
+        if (version !== missing) { // found
+          break;
+        }
+      } else {
+        if (!await fs.exists(manifestLoc)) {
+          pkgVersions.set(manifestLoc, missing);
+        } else if (!await fs.exists(path.join(manifestLoc, 'package.json'))) {
+          pkgVersions.set(manifestLoc, anyVersion);
+        } else {
+          ({version, dependencies} = await config.readManifest(manifestLoc, registryName));
+          pkgVersions.set(manifestLoc, version);
+          break;
+        }
+      }
+      searchPath.pop();
```

Comment:
Kind of hard to follow this logic. May be can use some comments. Also do we really need all this refactoring? If not let's defer that to a follow-up diff to see exactly what logic changes are needed to make this patch work.